sync/syncer.go

```diff
@@ -226,49 +223,47 @@ func (s *Syncer) getLayerBlockIDs(index types.LayerID) (chan types.BlockID, erro
 }
 
 func (s *Syncer) getIdsForHash(m map[string]p2p.Peer, index types.LayerID) (chan types.BlockID, error) {
-	reqCounter := 0
 
 	ch := make(chan []types.BlockID, len(m))
-	wg := sync.WaitGroup{}
-	for _, v := range m {
-		c, err := s.sendLayerBlockIDsRequest(v, index)
+	kill := make(chan struct{})
+	for _, peer := range m {
+		c, err := s.sendLayerBlockIDsRequest(peer, index)
 		if err != nil {
-			s.Error("could not fetch layer ", index, " block ids from peer ", v) // todo recover from this
+			s.Error("could not send layer %v ids ", index, " from peer ", peer) // todo recover from this
 			continue
 		}
-		reqCounter++
-		wg.Add(1)
-		go func() {
-			if v := <-c; v != nil {
-				ch <- v
+		go func(p p2p.Peer) {
+			select {
+			case v := <-c:
+				if v != nil {
+					ch <- v
+				}
+			case <-kill:
+				s.Info("ids request to %v timed out", p)
+				return
 			}
-			wg.Done()
-		}()
+		}(peer)
 	}
 
-	go func() { wg.Wait(); close(ch) }()
 	idSet := make(map[types.BlockID]bool, s.LayerSize)
-	timeout := time.After(s.RequestTimeout)
-	for reqCounter > 0 {
-		select {
-		case b := <-ch:
-			for _, id := range b {
-				bid := types.BlockID(id)
-				if _, exists := idSet[bid]; !exists {
-					idSet[bid] = true
-				}
-			}
-			reqCounter--
-		case <-timeout:
-			if len(idSet) > 0 {
-				s.Error("not all peers responded to hash request")
-				return keysAsChan(idSet), nil
-			}
-			return nil, errors.New("could not get block ids from any peer")
 
+	go func() {
+		<-time.After(s.RequestTimeout)
+		s.Info("timeout, not all peers responded to ids request")
+		close(kill)
+		close(ch)
+	}()
+
+	for _, bid := range <-ch {
```

Comment:
you can use for { select {} } and join the above goroutine here

Reply:
i know the select has its own disadvanteges 
im keeping it this way in the meantime until i find a better pattern